### ISSUE TITLE
Issue while converting a model with GRU to .tflite

### ISSUE THREAD

**Issue by ckmadhira** [tensorflow/tensorflow#97941]
### 1. System information

- Linux Ubuntu 22.04
- TensorFlow installation - pip package
- TensorFlow library - version 2.12.0

### 2. Code

GRU_model.py file attached
```

### 3. Failure after conversion
After generating the network, I see that a while loop in place of GRU. I tried to run the network with open source TFLM but, I am not able to. When I had logs in TFLM, it seems the GRU is actually decomposed into multiple operators but still not able to run the network.
When the GRU addition is replaced as below by having additional argument "unroll=True" then, I don't see any while loop in the tflite model. The GRU  is decomposed to basic operations.  But because stateful is set to True, I could see Quantization and dequantization operations and some additional floating point operations. When "stateful=False", I could see all the operators with int8 quantization. There are no float32 operators. What exactly stateful and unroll arguments are doing here? Does unroll is unrolling the time sequence of GRU or it also decomposing the GRU? why when stateful is set to False, I dont see any floating point operations. I assume that stateful is to feed the state variables to the next batch. How is it impacting the quantization and dequantization?

    x = tf.keras.layers.GRU(32, return_sequences=True, stateful=True, name='gruL', unroll=True)(x)

**Comment by ckmadhira** [comment]
[GRU_model.txt](https://github.com/user-attachments/files/21526863/GRU_model.txt)
Not able to upload GRU_model.py, so converted and uploaded a text file

**Comment by gaikwadrahul8** [comment]
Hi, @ckmadhira 
Thank you for bringing this issue to our attention, As far I know The unroll and stateful arguments in a Keras GRU layer have distinct effects on the generated TFLite model particularly impacting its control flow and quantization behavior.

The unroll argument modifies how a sequence is processed within a single inference call. By default (unroll=False) Keras uses a `tf.while_loop` which appears as a WHILE op in TFLite. This is efficient for handling sequences of varying lengths. When you set unroll=True Keras eliminates this loop and instead replicates the GRU's fundamental operations (matrix multiplications, additions) for each time step in the sequence. This creates a larger, static graph without a WHILE op which can be faster for short, fixed-length sequences but increases model size. The stateful argument controls how the GRU's hidden state is managed between separate inference calls. This is the critical factor impacting quantization.

With stateful=True option the model must preserve its internal hidden state (a float tensor) from one call to the next. During INT8 quantization the TFLite converter must handle this float state within an otherwise integer-based graph. To do so, it inserts Quantize and Dequantize operations around the state related calculations. This is why you observe a mixed-precision model with some float32 operations remaining they are necessary to maintain the state's precision and continuity across inferences.

On the other hand with `stateful=False` the GRU's state is reset for each new input. Since there is no persistent state to manage between calls the quantization process is greatly simplified. The TFLite converter can transform the entire model into a pure INT8 graph as there's no need to dequantize to float32 for state management. This results in a fully quantized often more performant model.

If full integer quantization proves challenging please consider alternative post-training quantization methods like `float16` quantization for reduced model size with minimal accuracy loss or `dynamic range quantization` which quantizes weights to int8 while activations are dynamically quantized at inference time often suitable for RNNs.

If I have missed something here please let me know. Thank you for your cooperation and understanding.

**Comment by ckmadhira** [comment]
Hi Rahul,
Thanks for the detailed explanation. 
One question I have here is, when unroll is set to false, the network will have while() operator. Why TFLM is not able to run this model? When I debugged it, I saw the concatenate() operator is getting different dimension (not at axis). Does TFLM supports GRU?

Thank you,
Chandrakala

**Comment by gaikwadrahul8** [comment]
Hi, @ckmadhira
TensorFlow Lite Micro (TFLM) does NOT support GRU operators when `unroll=False` because this configuration generates `while()` operators that are incompatible with TFLM's architecture. The concatenation dimension mismatch you're experiencing is a direct consequence of TFLM's inability to handle dynamic tensor shapes within while loop contexts. Please refer complete and current [list of supported operators](https://github.com/tensorflow/tflite-micro/blob/main/tensorflow/lite/micro/micro_mutable_op_resolver.h) and limitations of LiteRT for Microcontrollers (Previously known as TFLM) [official documentation ](https://ai.google.dev/edge/litert/microcontrollers/overview)

I would recommend you GRU layer to force unrolling by setting `unroll=True` or else go with supported LSTM Operators if that is feasible for your use case or project 

Thank you for your cooperation and understanding.
